labs/b9.md

```diff
@@ -0,0 +1,178 @@
+
+#### Puppet
+
+To get some hands on experience with Puppet, we are going to write a very basic manifest that pulls quotes from the internet and appends them to a file every two minutes. Although you were taught that in production environments we have the puppet master running on its own server, for simplicity you will apply the manifest locally, making you both the puppet master and the puppet agent. If you are confused about the Puppet vocabulary, check out slide 21.
+
+First things first. We plan to pull quotes from the web, so we need to ensure the `wget` package is installed on our system. Next we need to create a user, 'quotes', that runs this command for us. Take note of the dependency between the 'quotes' user and the 'quotegather' group existing. 
+
+Because we have full control over our user, let's make its home directory `/tmp`. Since this user has one purpose, to pull quotes from the web, they don't need a login shell. Go ahead and set it to `/bin/false`.
+
+Now we need a `cron` resource that grabs the quotes for us. I went ahead and filled in the command, so all you need to do is specify which user is to run the cron job and the interval at which it runs (every 2 minutes).
+
+Below I've provided a skeleton manifest `quotes.pp`. This can also be found in `~/decal-labs/b9/puppet/quotes.pp`, but comments are included here for your convenience.
+
+```puppet
+package { '': } # What package is required for this script to run?
+
+group { 'quotegather':
+  ensure  => '', # Do we want the group on this system?
+}
+
+user { 'quotes':
+  ensure  => '', # Do we want the user on this system?
+  gid     => , # Note that we can pass a group name to gid
+  home    => '', 
+  shell   => '',
+  # Do not create the user quotes unless the quotegather group exists
+  require => Group['quotegather'], 
+}
+
+cron { 'getquote':
+  user     => '',
+  # You can ignore this, it pulls the quotes from an API and appends
+  # them to a file.
+  command  => "curl -X POST http://api.forismatic.com/api/1.0/ -d 'method=getQuote' -d 'format=text' -d 'lang=en' >> quotes && echo >> quotes",
+  minute   => '', # Hint: passing * means this runs every minute
+}
+```
+Once you have completed your manifest, you can apply the changes to your system with the command `sudo puppet apply quotes.pp`. 
+
+Some debugging tips:
+- Read the [Offical Puppet Documentation](https://puppet.com/docs/puppet/5.5/type.html) about different resources if you are confused about commands
+- You can check your puppet syntax with `puppet parser validate quotes.pp`
+- List 'quotes' cron jobs with `sudo quotes -u quotes -l`
+
+Answer the below questions after you get the `quotes.pp` manifest working. Wait 10 minutes or so and running `cat /tmp/quotes` should yield a list of quotes.
+
+ ##### Questions
+
+ 1. Submit your completed `quotes.pp` file.
+ 2. What happens if you remove the line `require => Group['quotegather']` and run `sudo puppet apply quotes.pp` once more?
```

Comment:
Will this actually do anything? I feel like after the group is created already, it won't make a difference on future runs unless they delete the group and try again, and even then, puppet [runs in order for resources without dependencies](https://puppet.com/docs/puppet/4.8/lang_relationships.html#ordering), even in 4.8:

> By default, unrelated resources are managed in the order in which they’re written in their manifest file. If you declare an explicit relationship between resources, it will override this default ordering.

Reply:
Oy true. Good catch. >. <